Error in Matx::solve for non-square matrices

<!--
If you have a question rather than reporting a bug please go to http://answers.opencv.org where you get much faster responses.
If you need further assistance please read [How To Contribute](https://github.com/opencv/opencv/wiki/How_to_contribute).

This is a template helping you to create an issue which can be processed as quickly as possible. This is the bug reporting section for the OpenCV library.
-->

##### System information (version)
<!-- Example
- OpenCV => 3.1
- Operating System / Platform => Windows 64 Bit
- Compiler => Visual Studio 2015
-->

- OpenCV => latest


##### Detailed description
See the following code

https://github.com/opencv/opencv/blob/3165baa1f152878b49c984384aa61c430a686768/modules/core/include/opencv2/core/matx.hpp#L183

`solve` is supported for non-square `matx`, i.e., `m!=n`.

Its implementation is

https://github.com/opencv/opencv/blob/3165baa1f152878b49c984384aa61c430a686768/modules/core/include/opencv2/core/operations.hpp#L208-L222

Pay attention to the line
https://github.com/opencv/opencv/blob/3165baa1f152878b49c984384aa61c430a686768/modules/core/include/opencv2/core/operations.hpp#L214

`Matx_FastSolveOp` is defined as

https://github.com/opencv/opencv/blob/3165baa1f152878b49c984384aa61c430a686768/modules/core/include/opencv2/core/operations.hpp#L121-L133

which requires a square matrix `const Matx<_Tp, m, m>& a`


<!-- your description -->

##### Steps to reproduce

The following code will result in a compilation error

```.cpp
    cv::Matx<float, 3, 2> A(
            2, 1,
            3, 1,
            9, 1
    );
    cv::Vec<float, 3> b(4, 5, 7);

    auto x = A.solve(b, cv::DecompTypes::DECOMP_QR);
```

Error log is as follows:

```
........../include/opencv2/core/operations.hpp: In instantiation of ‘cv::Matx<_Tp, n, l> cv::Matx<_Tp, m, n>::solve(const cv::Matx<_Tp, m, l>&, int) const [with int l = 1; _Tp = float; int m = 3; int n = 2]’:
........../include/opencv2/core/matx.hpp:895:30:   required from ‘cv::Vec<_Tp, n> cv::Matx<_Tp, m, n>::solve(const cv::Vec<_Tp, m>&, int) const [with _Tp = float; int m = 3; int n = 2]’

........../include/opencv2/core/operations.hpp:214:57: error: no match for call to ‘(cv::internal::Matx_FastSolveOp<float, 3, 1>) (const cv::Matx<float, 3, 2>&, const cv::Matx<float, 3, 1>&, cv::Matx<float, 2, 1>&, int&)’
         ok = cv::internal::Matx_FastSolveOp<_Tp, m, l>()(*this, rhs, x, method);
                                                         ^
........../include/opencv2/core/operations.hpp:152:10: note: candidate: bool cv::internal::Matx_FastSolveOp<_Tp, 3, 1>::operator()(const cv::Matx<_Tp, 3, 3>&, const cv::Matx<T2, 3, 1>&, cv::Matx<T2, 3, 1>&, int) const [with _Tp = float]
     bool operator()(const Matx<_Tp, 3, 3>& a, const Matx<_Tp, 3, 1>& b,
          ^
........../include/opencv2/core/operations.hpp:152:10: note:   no known conversion for argument 1 from ‘const cv::Matx<float, 3, 2>’ to ‘const cv::Matx<float, 3, 3>&’
```

A workaround is to use
https://github.com/opencv/opencv/blob/3165baa1f152878b49c984384aa61c430a686768/modules/core/include/opencv2/core/operations.hpp#L215-L219


<!-- to add code example fence it with triple backticks and optional file extension
    ```.cpp
    // C++ code example
    ```
 or attach as .txt or .zip file
-->

--- a/modules/core/include/opencv2/core/operations.hpp
+++ b/modules/core/include/opencv2/core/operations.hpp
@@ -61,7 +61,16 @@ namespace cv
 namespace internal
 {
 
-template<typename _Tp, int m> struct Matx_FastInvOp
+template<typename _Tp, int m, int n> struct Matx_FastInvOp
+{
+    bool operator()(const Matx<_Tp, m, n>&, Matx<_Tp, n, m>&, int) const
+    {
+        CV_Assert(false);
+        return false;
+    }
+};
+
+template<typename _Tp, int m> struct Matx_FastInvOp<_Tp, m, m>
 {
     bool operator()(const Matx<_Tp, m, m>& a, Matx<_Tp, m, m>& b, int method) const
     {
@@ -78,7 +87,7 @@ template<typename _Tp, int m> struct Matx_FastInvOp
     }
 };
 
-template<typename _Tp> struct Matx_FastInvOp<_Tp, 2>
+template<typename _Tp> struct Matx_FastInvOp<_Tp, 2, 2>
 {
     bool operator()(const Matx<_Tp, 2, 2>& a, Matx<_Tp, 2, 2>& b, int) const
     {
@@ -94,7 +103,7 @@ template<typename _Tp> struct Matx_FastInvOp<_Tp, 2>
     }
 };
 
-template<typename _Tp> struct Matx_FastInvOp<_Tp, 3>
+template<typename _Tp> struct Matx_FastInvOp<_Tp, 3, 3>
 {
     bool operator()(const Matx<_Tp, 3, 3>& a, Matx<_Tp, 3, 3>& b, int) const
     {
@@ -118,7 +127,17 @@ template<typename _Tp> struct Matx_FastInvOp<_Tp, 3>
 };
 
 
-template<typename _Tp, int m, int n> struct Matx_FastSolveOp
+template<typename _Tp, int m, int l, int n> struct Matx_FastSolveOp
+{
+    bool operator()(const Matx<_Tp, m, l>&, const Matx<_Tp, m, n>&,
+                    Matx<_Tp, l, n>&, int) const
+    {
+        CV_Assert(false);
+        return false;
+    }
+};
+
+template<typename _Tp, int m, int n> struct Matx_FastSolveOp<_Tp, m, m, n>
 {
     bool operator()(const Matx<_Tp, m, m>& a, const Matx<_Tp, m, n>& b,
                     Matx<_Tp, m, n>& x, int method) const
@@ -132,7 +151,7 @@ template<typename _Tp, int m, int n> struct Matx_FastSolveOp
     }
 };
 
-template<typename _Tp> struct Matx_FastSolveOp<_Tp, 2, 1>
+template<typename _Tp> struct Matx_FastSolveOp<_Tp, 2, 2, 1>
 {
     bool operator()(const Matx<_Tp, 2, 2>& a, const Matx<_Tp, 2, 1>& b,
                     Matx<_Tp, 2, 1>& x, int) const
@@ -147,7 +166,7 @@ template<typename _Tp> struct Matx_FastSolveOp<_Tp, 2, 1>
     }
 };
 
-template<typename _Tp> struct Matx_FastSolveOp<_Tp, 3, 1>
+template<typename _Tp> struct Matx_FastSolveOp<_Tp, 3, 3, 1>
 {
     bool operator()(const Matx<_Tp, 3, 3>& a, const Matx<_Tp, 3, 1>& b,
                     Matx<_Tp, 3, 1>& x, int) const
@@ -195,7 +214,7 @@ Matx<_Tp, n, m> Matx<_Tp, m, n>::inv(int method, bool *p_is_ok /*= NULL*/) const
     Matx<_Tp, n, m> b;
     bool ok;
     if( m == n && (method == DECOMP_LU || method == DECOMP_CHOLESKY) )
-        ok = cv::internal::Matx_FastInvOp<_Tp, m>()(*reinterpret_cast<const Matx<_Tp, m, m>*>(this), reinterpret_cast<Matx<_Tp, m, m>&>(b), method);
+        ok = cv::internal::Matx_FastInvOp<_Tp, m, n>()(*this, b, method);
     else
     {
         Mat A(*this, false), B(b, false);
@@ -211,7 +230,7 @@ Matx<_Tp, n, l> Matx<_Tp, m, n>::solve(const Matx<_Tp, m, l>& rhs, int method) c
     Matx<_Tp, n, l> x;
     bool ok;
     if( m == n && (method == DECOMP_LU || method == DECOMP_CHOLESKY) )
-        ok = cv::internal::Matx_FastSolveOp<_Tp, m, l>()(*reinterpret_cast<const Matx<_Tp, m, m>*>(this), rhs, reinterpret_cast<Matx<_Tp, m, l>&>(x), method);
+        ok = cv::internal::Matx_FastSolveOp<_Tp, m, n, l>()(*this, rhs, x, method);
     else
     {
         Mat A(*this, false), B(rhs, false), X(x, false);

--- a/modules/core/include/opencv2/core/operations.hpp
+++ b/modules/core/include/opencv2/core/operations.hpp
@@ -210,8 +210,8 @@ Matx<_Tp, n, l> Matx<_Tp, m, n>::solve(const Matx<_Tp, m, l>& rhs, int method) c
 {
     Matx<_Tp, n, l> x;
     bool ok;
-    if( method == DECOMP_LU || method == DECOMP_CHOLESKY )
-        ok = cv::internal::Matx_FastSolveOp<_Tp, m, l>()(*this, rhs, x, method);
+    if( m == n && (method == DECOMP_LU || method == DECOMP_CHOLESKY) )
+        ok = cv::internal::Matx_FastSolveOp<_Tp, m, l>()(*reinterpret_cast<const Matx<_Tp, m, m>*>(this), rhs, reinterpret_cast<Matx<_Tp, m, l>&>(x), method);
     else
     {
         Mat A(*this, false), B(rhs, false), X(x, false);

--- a/modules/core/test/test_math.cpp
+++ b/modules/core/test/test_math.cpp
@@ -3129,6 +3129,21 @@ TEST(Core_QR_Solver, accuracy64f)
     ASSERT_FALSE(solve(A, B, solutionQR, DECOMP_QR));
 }
 
+TEST(Core_Solve, regression_11888)
+{
+    cv::Matx<float, 3, 2> A(
+        2, 1,
+        3, 1,
+        6, 1
+    );
+    cv::Vec<float, 3> b(4, 5, 7);
+    cv::Matx<float, 2, 1> xQR = A.solve(b, DECOMP_QR);
+    cv::Matx<float, 2, 1> xSVD = A.solve(b, DECOMP_SVD);
+    EXPECT_LE(cvtest::norm(xQR, xSVD, CV_RELATIVE_L2), FLT_EPSILON);
+    cv::Matx<float, 2, 3> iA = A.inv(DECOMP_SVD);
+    EXPECT_LE(cvtest::norm(A*iA, Matx<float, 3, 3>::eye(), CV_RELATIVE_L2), 0.6);
+}
+
 softdouble naiveExp(softdouble x)
 {
     int exponent = x.getExp();

